src/module.c

```diff
@@ -675,14 +630,29 @@ void moduleFreeContext(RedisModuleCtx *ctx) {
     if (ctx->flags & REDISMODULE_CTX_THREAD_SAFE) freeClient(ctx->client);
 }
 
+/* Create a module ctx and keep track of the nesting level.
+ *
+ * Note: When creating ctx for threads (RM_GetThreadSafeContext and
+ * RM_GetDetachedThreadSafeContext) we do not bump up the nesting level
+ * because we only need to track of nesting level in the main thread
+ * (only the main thread uses propagatePendingCommands) */
+void moduleCreateContext(RedisModuleCtx *out_ctx, RedisModule *module, int ctx_flags) {
+    memset(out_ctx, 0 ,sizeof(RedisModuleCtx));
+    out_ctx->getapifuncptr = (void*)(unsigned long)&RM_GetApi;
+    out_ctx->module = module;
+    out_ctx->flags = ctx_flags;
+    if (out_ctx->module && !(ctx_flags & REDISMODULE_CTX_THREAD_SAFE)) {
+        out_ctx->module->ctx_nesting++;
```

Comment:
comment from @MeirShpilraien 
ctx_nesting should be global, not per module
imagine "module1.foo" calling "module2.foo". if the nesting level is not global module2.foo will propagate everything by itself, while module1.foo should do it (the issue here is the order of propagation)